sdks/go/pkg/beam/core/runtime/harness/harness.go

```diff
@@ -203,20 +204,44 @@ func (c *control) handleInstruction(ctx context.Context, req *fnpb.InstructionRe
 
 		err := plan.Execute(ctx, id, c.data)
 
+		snapshot := plan.ProgressReport()
+		planID := plan.ID()
 		// Move the plan back to the candidate state
 		c.mu.Lock()
 		c.plans[plan.ID()] = plan
 		delete(c.active, id)
 		c.mu.Unlock()
 
+		transforms := map[string]*fnpb.Metrics_PTransform{
```

Comment:
Could plan.ProgressReport return transforms instead? Then we'd also not need to expose the ParDoIDs.

Reply:
The ProgressReport is forwarded from the Datasource, which is also the file that defines the ProgressReport. It would be weird to have a "sidecar" field defined in a file that doesn't know or care about those ParDoIDs only to have it populated if the ProgressReport method is called from something else. It feels less maintainable than the current implementation, which leverages that we know a Plan is that type, rather than by an interface type, allowing access to exported fields.
Having an exported field is preferable to a Getter function in Go IIRC, unless we are particularly concerned about abuse of these opaque per bundle PTransform identifiers leaking out.

Reply:
We could add another method that returns the Metrics (incl. progress report) to achieve the same thing, if we call ProgressReport from multiple places. That would keep the ParDo IDs and the code that uses it together, which I think will make the code easier to follow.